### PR TITLE
NDRS-857/sample memory metrics in consensus

### PR DESCRIPTION
**Pull request by marc-casperlabs**
Reduces heap measuring of large collections by sampling only a subset of their contents.

### CODE REVIEW

File: node/src/utils/ds.rs
Comment (afck) [inline node/src/utils/ds.rs:25]: The `size_of::<usize>()` was probably copied from `Vec` — for hash sets and maps it's more complicated, but maybe the overhead is negligible?

File: node/src/utils/ds.rs
Comment (afck) [inline node/src/utils/ds.rs:39]:
```suggestion
    // Initialize a buffer suitable for the seed, which might be larger than our length bytes.
```

File: node/src/utils/ds.rs
Comment (afck) [inline node/src/utils/ds.rs:57]:
```suggestion
/// Extrapolate memory usage of a `Vec` by from a random subset of `SAMPLE_SIZE` items.
```

File: node/src/utils/ds.rs
Comment (afck) [inline node/src/utils/ds.rs:75]:
```suggestion
/// Extrapolate memory usage of a `HashMap` by from a random subset of `SAMPLE_SIZE` items.
```

File: node/src/utils/ds.rs
Comment (afck) [inline node/src/utils/ds.rs:69]: As discussed, let's make sure this estimates the `T` and not `&T`. (Also below.)

**Review by afck — COMMENTED** [review]
As discussed I reviewed only https://github.com/CasperLabs/casper-node/pull/971/commits/ba270b3a0e0470e3aa3cac10b6c8e6e643644b91 and https://github.com/CasperLabs/casper-node/pull/971/commits/68561678c30448638429f0c2bc9a5b7a5d27a5aa. Looks good to me so far, but let's make sure we're not estimating `&T` instead of `T`. (See my comment.)

File: node/src/utils/ds.rs
Comment (marc-casperlabs) [reply node/src/utils/ds.rs:25]: It has been copied from the estimate from `datasize` itself, unfortunately I did not look in details into the accuracy of these metrics, as that would involve very intricate knowledge of how those data structures are implemented.

Overall this is why `datasize` says "estimate" on the label.

File: node/src/utils/ds.rs
Comment (marc-casperlabs) [reply node/src/utils/ds.rs:69]: Can do, I will add a unit test for this.

**Review by afck — APPROVED** [review]
Approving, assuming https://github.com/CasperLabs/casper-node/pull/969 gets approved and the `T` vs. `&T` tested.

File: node/src/utils/ds.rs
Comment (afck) [inline node/src/utils/ds.rs:72]: I thought that was exactly equivalent to `.map(DataSize::estimate_heap_size)`?? Don't you have to `(*v).estimate_heap_size()` or something?

File: node/src/utils/ds.rs
Comment (afck) [inline node/src/utils/ds.rs:66]: Right, good catch!

**Review by afck — APPROVED** [review]

File: node/src/utils/ds.rs
Comment (goral09) [inline node/src/utils/ds.rs:161]: Isn't the point of this "sampling" functions to provide a _good enough_ estimate where it's expensive to calculate the exact value? That's not the case for primitive types though, right?

Comment (afck) [reply node/src/utils/ds.rs:161]: The exact calculation would iterate over all entries in the hash map, and sum up their sizes, whereas the estimate only takes the average of a sample, and multiplies it by the number of entries.

Even for a hash map of vectors, that should be faster than the full iteration.

Comment (goral09) [reply node/src/utils/ds.rs:161]: I get that. My point is that this particular test is not very useful in a sense that it doesn't test what's been added here. It doesn't verify how many iterations we did so `hashmap_sample` could be doing exactly what you described.

Comment (afck) [reply node/src/utils/ds.rs:161]: Yes, I think that test is just to verify that the new method is correct, not that it's faster. But I'll let @marc-casperlabs comment on that.

Comment (goral09) [reply node/src/utils/ds.rs:161]: And it's trivial for the primitive types since its size is constant.

Comment (afck) [reply node/src/utils/ds.rs:161]: But in this case it's not trivial because it's a map of _vectors_ of `u32`s, not a map of `u32`s.

**Pull request merged**
